### ISSUE TITLE
broken navigation on FF 54 with synchronous ajax calls on unload

### ISSUE THREAD

**Issue by adimascio** [EFForg/privacybadger#1438]
With firefox 54, there seems to be a conflict between synchronous
ajax queries performed on `window.unload` and Privacy Badger. When
the user clicks on a link, the url changes in the URL navigation bar
but the page content doesn't update. The user has to reload the page
manually.

Although synchronous ajax queries are a very bad idea, and hooking
on `unload` event is probably bad too, it happens to break navigation 
on some websites (e.g. https://www.logilab.org/).

**NOTE: This is specific to firefox**

This bug was observed in the following environment:

- Privacy Badger version: 2017.6.13.1
- Firefox version: 54.0 (64 bits)
- Debian package version: 54.0-1
- Debian Version: Stretch

Steps to reproduce the bug (using a [minimal test website](https://github.com/adimascio/testff)):
```
 $ git clone https://github.com/adimascio/testff.git
 $ cd testff
 $ npm install
 $ node index.js
 ```

Now, open your browser on `http://locahost:3000` and try to follow `page1` and `page2` links.

I'm willing to provide a PR but I'll need some help to know where to start digging in Privacy Badger code base.

**Comment by ghostwords** [comment]
Sounds related to https://github.com/EFForg/privacybadger/issues/1357#issuecomment-298348827, https://github.com/EFForg/privacybadger/issues/1295, https://github.com/EFForg/privacybadger/issues/1266.

**Comment by cowlicks** [comment]
Where on the logilab site were you navigating when you saw this? I was not able to reproduce this there.

**Comment by cowlicks** [comment]
I think issues like this are a good use case for https://github.com/EFForg/privacybadger/pull/1221

**Comment by adimascio** [comment]
@ghostwords this seems very similar indeed

**Comment by adimascio** [comment]
@cowlicks after I disabled all extensions but PrivacyBadger, I can't navigate on any page of logilab.org without reloading the page after a click on a link.

I might not have been clear in the issue description but `[the small github repository](https://github.com/adimascio/testff.git)` I mentioned there is the tiniest test program I could come up with to reproduce the bug.

**Comment by cowlicks** [comment]
@adimascio I'll try it out thanks!

**Comment by adimascio** [comment]
@cowlicks great, thanks !

**Comment by cowlicks** [comment]
@adimascio I think the "`server.js`" file is missing.

**Comment by cowlicks** [comment]
ah I see you meant "`index.js`"

**Comment by adimascio** [comment]
You're right, sorry. I should have proofread my README :-/

**Comment by ghostwords** [comment]
I made a minimal extension to reproduce the issue. @adimascio Could you double-check that you run into the bug using just the attached add-on and nothing else? It would be good to know affected Firefox versions as well.

[firefox_webext_onunload_bug_demo.zip](https://github.com/EFForg/privacybadger/files/1079120/firefox_webext_onunload_bug_demo.zip)

The next step would be to file a Mozilla Bugzilla bug.

**Comment by ghostwords** [comment]
To load the above add-on in Firefox, unzip the archive somewhere, visit `about:debugging`, click on Load Temporary Add-on, and pick the `manifest.json` file inside the unzipped folder.

**Comment by ghostwords** [comment]
>I think issues like this are a good use case for #1221 

I dunno, I don't see any helpful messages. Everything seems fine, it's just that Firefox fails to render the newly navigated-to page.

**Comment by adimascio** [comment]
@ghostwords Thanks ! I confirm that I run into the bug after loading your extension (I created a new firefox profile with no other extension).

**Comment by adimascio** [comment]
@ghostwords I've just made a few more tests with your extension. It looks like returning `undefined` or explicitly setting a `responseHeaders` property with at least `content-type` and `content-length` solves the problem.

**Comment by adimascio** [comment]
Seems like https://github.com/adimascio/privacybadger/commit/8483d8c249ced9ba8eb7567310769a59fe48eb4f might fix the problem. I will try to double-check later today

**Comment by ghostwords** [comment]
Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1373776.

**Comment by naught101** [comment]
That zip file link is not working for me. Returns an xml file with an error from amazon AWS complaining about "The request signature we calculated does not match the signature you provided. Check your key and signing method." when I try to click the link.

**Comment by ghostwords** [comment]
[This one](https://github.com/EFForg/privacybadger/files/1079120/firefox_webext_onunload_bug_demo.zip)? Is it still an issue? Just worked for me.

**Comment by naught101** [comment]
Ahh.. NoScript was blocking it as and XSS attempt. Got it now.

**Comment by naught101** [comment]
Yep, that debug addon does cause the same problem as Privacy Badger for me.

**Comment by ghostwords** [comment]
According to https://github.com/EFForg/privacybadger/issues/1295#issuecomment-315998618, you should be able to get rid of this issue in your Firefox profile without having to start over from scratch by [refreshing your Firefox](https://support.mozilla.org/en-US/kb/refresh-firefox-reset-add-ons-and-settings) and then reinstalling all your add-ons.

**Comment by naught101** [comment]
Browser refresh worked for ebay for me, but not for my internet banking login screen, for some reason (https://ibanking.stgeorge.com.au/ibank/loginPage.action). Not sure why... Let me know if there is more testing I can do.

**Comment by koops76** [comment]
@adimascio File a bug on Mozilla tracker.

**Comment by ghostwords** [comment]
This might be fixed now that Firefox 55 reached stable.

**Comment by arthurlogilab** [comment]
@ghostwords unfortunately I can confirm that this bug still present on firefox 55

**Comment by ghostwords** [comment]
@arthurlogilab Are you using the latest stable (not beta) Firefox 55 version? Could you chime in on the Firefox issue, following up on my findings there (https://bugzilla.mozilla.org/show_bug.cgi?id=1373776#c18)?

**Comment by arthurlogilab** [comment]
Sorry, I take it back, it is solved on my firefox 55.0.1 on ubuntu zesty version. 🍾
